src/components/SetupTab/index.jsx

```diff
@@ -2,41 +2,354 @@ import React from 'react';
+  * in the 'ui_control' array of `argsSpec[argkey]`.
+  *
+  * @params {object} argsSpec - merge of an InVEST model's ARGS_SPEC and UI Spec.
+  * @params {object} argsValues - of the shape returned by `initializeArgValues`.
+  * @params {string} argkey - a key of the argsSpec and argsValues objects
+  *    that contains a 'ui_control' property.
+  * 
+  * @return {object} a copy of `argsValues`
+  */
+  let updatedValues = Object.assign({}, argsValues)
+  argsSpec[argkey].ui_control.forEach(dependentKey => {
+    if (!updatedValues[argkey].value) {
+      // apply the display option specified in the UI spec
+      updatedValues[dependentKey]['ui_option'] = argsSpec[dependentKey].ui_option
+    } else {
+      updatedValues[dependentKey]['ui_option'] = undefined
+    }
+  });
+  return(updatedValues)
+}
 
-import { fetchDatastackFromFile } from '../../server_requests';
-import { argsValuesFromSpec } from '../../utils';
+function initializeArgValues(argsSpec, argsDict) {
+  /** Setup the objects that store InVEST argument values in SetupTab state.
+  *
+  * One object will store input form values and track if the input has been
+  * touched. The other object stores data returned by invest validation.
+  * 
+  * 
+  * @params {object} argsSpec - merge of an InVEST model's ARGS_SPEC.args and UI Spec.
+  * @params {object} argsDict - key: value pairs of InVEST model arguments, or {}.
+  * 
+  * @return {object} to destructure into two args, 
+  *   each with the same keys as argsSpec:
+  *     {object} argsValues - stores properties that update in response to
+  *       user interaction
+  *     {object} argsValidation - stores properties that update in response to
+  *       validation.
+  */
+  const initIsEmpty = Object.keys(argsDict).length === 0
+  let argsValidation = {};
+  let argsValues = {};
+  Object.keys(argsSpec).forEach((argkey) => {
+    argsValidation[argkey] = {}
+    if (argkey === 'n_workers') { return }
+    argsValues[argkey] = {
+      value: argsDict[argkey],
+      touched: !initIsEmpty  // touch them only if initializing with values
+    }
+  });
+  return({ argsValues: argsValues, argsValidation: argsValidation })
+}
 
 export class SetupTab extends React.Component {
-  /** Renders an Arguments form and an Execute button
+  /** Renders an arguments form, execute button, and save button.
   */
+  constructor(props) {
+    super(props);
+    this.state = {
+      argsValues: null,
+      argsValidation: {},
+      argsValid: false,
+      sortedArgTree: null
+    }
+
+    this.savePythonScript = this.savePythonScript.bind(this);
+    this.wrapArgsToJsonFile = this.wrapArgsToJsonFile.bind(this);
+    this.wrapInvestExecute = this.wrapInvestExecute.bind(this);
+    this.investValidate = this.investValidate.bind(this);
+    this.updateArgValues = this.updateArgValues.bind(this);
+    this.batchUpdateArgs = this.batchUpdateArgs.bind(this);
+  }
+
+  componentDidMount() {

```

Comment:
Recommended reading: https://reactjs.org/blog/2018/06/07/you-probably-dont-need-derived-state.html